### PR TITLE
Use FeatureDeprecated (and other cleanups)

### PR DESCRIPTION
**Pull request by dcbaker**
This little series was meant to mainly replace a number of `mlog.deprecation` calls with `FeatureDeprecated`, along the way I did a little cleanup.

Suggested by @mensinda

### CODE REVIEW

File: mesonbuild/optinterpreter.py [hunk 1]
```diff
@@ -200,11 +201,8 @@ def evaluate_statement(self, node):
             raise OptionException('Only calls to option() are allowed in option files.')
         (posargs, kwargs) = self.reduce_arguments(node.args)
 
-        # FIXME: Cannot use FeatureNew while parsing options because we parse
-        # it before reading options in project(). See func_project() in
-        # interpreter.py
-        #if 'yield' in kwargs:
-        #    FeatureNew('option yield', '0.45.0').use(self.subproject)
```
Comment (mensinda) [inline mesonbuild/optinterpreter.py:207]: There should be an assertion here that `project_meson_versions` is set, since assumptions about the initialization order break easily during a refactoring (especially if there are no comments).

Comment (dcbaker) [reply mesonbuild/optinterpreter.py:207]: I've done the best I can to address this in the latest version of the series. There is a corner case that's in our test suite that makes it kinda hard to solve generally, so the approach I have is a bit of a hack (I really don't want to rewrite the way this detection works completely), that corner is if a FeatureNew/FeatureDeprecated is triggered in an argument to project(), such as
```meson
project('foo', version : run_command(..., check : true))
```
So I've changed the type of project_meson_versions to a defaultdict so there is always a value, and added a comment to project() implementation

Comment (mensinda) [reply mesonbuild/optinterpreter.py:207]: Ah, I didn't see this corner case immediately...

**Comment by jon-turney** [comment]
It looks like after this test case `warning/1 version for string div` sometimes doesn't output the expected  `tried to use feature introduced in '0.49.0': / with string arguments`? No idea why that might be!
